src/utils/wallet.js

```diff
@@ -438,94 +521,134 @@ class Wallet {
+            return
+        }
+        // create account if new
+        if (isNew) {
+            await this.createNewAccount(accountId, fundingContract, fundingKey)
+        }
+        // now finish recovery method setup
         const { seedPhrase, publicKey } = generateSeedPhrase();
-
-        const account = this.getAccount(accountId)
+        const { account, has2fa } = await this.getAccountAndState(accountId)
         const accountKeys = await account.getAccessKeys();
-        if (!accountKeys.some(it => it.public_key.endsWith(publicKey))) {
-            await account.addKey(publicKey);
+        if (has2fa) {
+            await this.addAccessKey(account.accountId, account.accountId, convertPKForContract(publicKey))
+        } else {
+            if (!accountKeys.some(it => it.public_key.endsWith(publicKey))) {
+                await account.addKey(publicKey);
+            }
         }
-
         return sendJson('POST', `${ACCOUNT_HELPER_URL}/account/sendRecoveryMessage`, {
             accountId,
             method,
             seedPhrase
         });
     }
 
-    async replaceAccessKey(oldKey, newKey) {
-        const accountId = this.accountId;
-        await this.getAccount(accountId).addKey(newKey)
-        await this.removeAccessKey(oldKey)
-    }
-
     async sendNewRecoveryLink(method) {
         const accountId = this.accountId;
-        const { seedPhrase, publicKey } = generateSeedPhrase();
+        const { account, has2fa } = await this.getAccountAndState(accountId)
+        const { seedPhrase, publicKey } = generateSeedPhrase()
+
+        if (has2fa) {
+            await this.twoFactor.rotateKeys(account, publicKey, method.publicKey)
+        } else {
+            await account.addKey(publicKey)
+            await this.removeAccessKey(method.publicKey)
+        }
 
-        await this.postSignedJson('/account/resendRecoveryLink', {
+        return await this.postSignedJson('/account/resendRecoveryLink', {
             accountId,
             method,
             seedPhrase,
             publicKey
         });
-        await this.replaceAccessKey(method.publicKey, publicKey)
+
     }
 
-    async deleteRecoveryMethod({ kind, publicKey }) {
-        await this.postSignedJson('/account/deleteRecoveryMethod', {
-            accountId: this.accountId,
-            kind,
-            publicKey
-        })
-        await this.removeAccessKey(publicKey)
+    async deleteRecoveryMethod({ kind, publicKey }, deleteAllowed = true) {
+        const accessKeys =  await this.getAccessKeys()
+        const pubKeys = accessKeys.map(key => key.public_key)
+
+        if (deleteAllowed) {
+            if (pubKeys.includes(publicKey)) {
+                await this.removeAccessKey(publicKey)
+            }
+            await this.postSignedJson('/account/deleteRecoveryMethod', {
+                accountId: this.accountId,
+                kind,
+                publicKey
+            })
+        } else {
```

Comment:
Seems like this should just be done by caller. Why pass parameter here just to forcing throwing error?